### PR TITLE
Swift: Data flow through optional binding

### PR DESCRIPTION
**Pull request by geoffw0**
Data flow through optional binding, that is:
```
if let x = [optional expression] {
  ...
```

First there are lots of tests, because I was nervous about edge cases.  It turns out `switch case` has the trickiest cases, and we don't support that yet, `if let` wasn't so bad (I intend to come back to `switch case` at a later date).

The change itself, in `DataFlowPrivate.qll`, is straightforward.  I also had to generalize a case in `Ssa.qll`s `variableWrite` predicate, allowing all variables declared in patterns to be Ssa variables, not just those with an associated `PatternBindingDecl` (i.e. a top level `let x = y` or `var x = y` statement).  I think this is reasonable but I'd appreciate @MathiasVP s view.

TODO:
 - [x] check DCA run.
 - [x] create follow-up issue for `switch case`.

### CODE REVIEW

**Comment by geoffw0** [comment]
Not much to see on the DCA run - looks good apart from unrelated database creation wobbles.

File: swift/ql/lib/codeql/swift/dataflow/Ssa.qll [hunk 1]
```diff
@@ -30,9 +30,16 @@ module Ssa {
...
+      // let x2 = v
+      // let (x3, x4) = tuple
+      // if let x5 = optional { ... }
+      // guard let x6 = optional else { ... }
+      // ```
+      exists(Pattern pattern |
         bb.getNode(i).getNode().asAstNode() = pattern and
-        decl.getAPattern() = pattern and
         v.getParentPattern() = pattern and
         certain = true
```
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/dataflow/Ssa.qll:44]: Indeed, I think this looks totally fine 👍.

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll [hunk 2]
```diff
@@ -153,6 +153,13 @@ private module Cached {
     or
     nodeFrom.asExpr() = nodeTo.asExpr().(OptionalEvaluationExpr).getSubExpr()
     or
+    // flow through optional binding `if let`
+    exists(ConditionElement ce, ConcreteVarDecl v |
+      ce.getInitializer() = nodeFrom.asExpr() and
+      ce.getPattern() = v.getParentPattern() and
+      nodeTo.asDefinition().getSourceVariable() = v
+    )
+    or
```
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:162]: Does this mean we have flow from `source()` to both `x` and `y` in something like:
```swift
func source() -> (Int, Int)? { ... }

if let (x, y) = source() {
  print(x, y)
}
```

I would have hoped that this case wasn't needed, but that it would be covered by the very first disjunct in this predicate (possibly by extending `assigns`):
```ql
def.(Ssa::WriteDefinition).assigns(nodeFrom.getCfgNode()) and
nodeTo.asDefinition() = def
```

**Review by MathiasVP — COMMENTED** [review]
One question. Otherwise, this LGTM.

File: swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll (hunk 2 above)
Comment (geoffw0) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:162]: I've added your example to the tests.

We do have flow from `source()` to both `x` and `y` here, provided by the `if let` case in `DataFlowPrivate.qll`.  Its needed for all the new results (except line 303) I think.  The SSA assignment case doesn't cover them, but perhaps it could???

Comment (MathiasVP) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:162]: Sorry if I wasn't clear. What I'm saying is two-fold:
1. I would not expect dataflow from `source()` to `x` and `y` as _dataflow_ step. If anything, it should be a _taint_ step.
2. I don't think the case added should be here. Instead, it should be part of the [first case in this predicate](https://github.com/github/codeql/blob/18a815ca8b5fb01d8867b188a4e5213e7d530154/swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll#L119), where `assigns` is extended to transfer flow from `a` to `b` when it's of the form `if let a = b`.

Does that make sense?

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll:162]: > I would not expect dataflow from source() to x and y as dataflow step. If anything, it should be a taint step.

Good point.  In general I've been happy that these are dataflow steps, but this case is an exception.

I'll take a look at your proposed fix (when I can find 15 minutes)...

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll]: ... I've updated the logic to use `Ssa::Definition` in place of `ConcreteVarDecl`, and then updated it again moving that into `Ssa::WriteDefinition.assigns` itself.  That feels like probably an improvement but I'm not very comfortable in the SSA library internals, and I'm not sure what else this change now affects.  I would appreciate your thoughts.

Regarding the `if let (x, y) = source()` case in particular, I think a full fix might involve decoding the `Pattern` on the left side and matching it against access paths of the thing on the right side of the assignment ... potentially a bit of a huge time sink.  I'm inclined to just exclude cases where more than one variable is initialized by the `ConditionElement`.  Or write a taint flow case for these cases perhaps.  Do you have any better ideas?

Comment (MathiasVP) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll]: > That feels like probably an improvement but I'm not very comfortable in the SSA library internals, and I'm not sure what else this change now affects. I would appreciate your thoughts.

I think what you've done in 2c35af51cd694a3456f6b8de4283b5f899df32da makes a lot of sense. IIRC, the `assigns` predicate is only used in `localFlowStepCommon`. So your change is still a very local one 👍.

> Regarding the if let (x, y) = source() case in particular, I think a full fix might involve decoding the Pattern on the left side and matching it against access paths of the thing on the right side of the assignment ... potentially a bit of a huge time sink.

Yeah, I completely agree. This proposed pattern decoding should already be covered by https://github.com/github/codeql/pull/11111, no? Of course, without a real implementation of `source()` that actually produces a tuple containing a tainted element, field flow won't kick in in the testcase you've added in https://github.com/github/codeql/pull/11836/commits/4e5483744f9124584dce02b346f847477c83a604 (which is totally fine!).

> I'm inclined to just exclude cases where more than one variable is initialized by the ConditionElement. Or write a taint flow case for these cases perhaps. Do you have any better ideas?

I think this is exactly the right approach, yes. We should have dataflow in the case where just one variable is initialized, but as soon as it's pattern that's producing more than one variable we shouldn't have dataflow.

I haven't fully grasped whether we want taintflow in this case. It will give us taint FPs on something like:
```swift
func second_is_tainted() -> (Int, Int)? {
  return (0, source())
}

...
if let (x, y) = second_is_tainted() {
  sink(x)
}
```

but, of course, it might actually give us some good taintflow in case a query is looking for something like:
```swift
func data_from_network() -> (Int, Int)? {
  string from_network = receive_data()
  return __internal_decode_data(from_network) // let's say we don't have the body of this function.
}

...
if let (x, y) = data_from_network() {
  sink(x)
}
```

Comment (geoffw0) [reply swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll]: > This proposed pattern decoding should already be covered by #11111, no? Of course, without a real implementation of source() that actually produces a tuple containing a tainted element, field flow won't kick in in the testcase you've added

I see what you mean about `source()` in this test case.  I'm not sure if content flow through tuples will need a special case for decoding `if let (x, y)` as well.  I can't imagine any of the existing cases quite cover it successfully (I could be wrong).  And your two examples above show that we will need to consider a variety of cases quite carefully when we do get around to solving these sorts of problems.

> We should have dataflow in the case where just one variable is initialized, but as soon as it's pattern that's producing more than one variable we shouldn't have dataflow.

I've added an exclusion (and a commit with a QL-for-QL fix after that).  I'm not convinced its the most accurate / concise way of expressing it so let me know if you can think of a better way.

File: swift/ql/lib/codeql/swift/dataflow/Ssa.qll [hunk 3]
```diff
@@ -169,7 +169,8 @@ module Ssa {
       exists(VarDecl var, SsaInput::BasicBlock bb, int blockIndex, ConditionElement ce, Expr init |
         this.definesAt(var, bb, blockIndex) and
         ce.getPattern() = bb.getNode(blockIndex).getNode().asAstNode() and
-        init = ce.getInitializer()
+        init = ce.getInitializer() and
+        strictcount(Ssa::WriteDefinition alt | alt.definesAt(_, bb, blockIndex)) = 1 // exclude cases where there are multiple writes from the same pattern, this is at best taint flow.
```
Comment (MathiasVP) [inline swift/ql/lib/codeql/swift/dataflow/Ssa.qll:173]: I imagine we could also write this by inspecting the structure of `ce.getPattern()`, but I appreciate the simplicity of this solution 👍.

**Review by MathiasVP — APPROVED** [review]
LGTM!
